refactor: Introduce GlobalQuery class: the plugin uses a singleton.

ilandikov opened the pull request:
# Description

~~Move `GlobalQuery` from `Settings` to a singleton abstraction.~~

Provide a `GlobalQuery` class to encapsulate a lot of the behaviour of global query setting.

It is based on ideas in the existing `StatusRegistry` class, where tests create instances of the class and use them in isolation. And code in `src/` only uses the static `GlobalQuery.getInstance()` to obtain a `GlobalQuery` object.

This PR is a refactored PR https://github.com/obsidian-tasks-group/obsidian-tasks/pull/2209

## Motivation and Context

We discussed several times that this is needed so I found some time =)

## How has this been tested?

Unit tests.

## Types of changes


Internal changes:

- [x] **Refactor** (prefix: `refactor` - non-breaking change which only improves the design or structure of existing code, and making no changes to its external behaviour)
- [x] **Tests** (prefix: `test` - additions and improvements to unit tests and the smoke tests)
- [ ] **Infrastructure** (prefix: `chore` - examples include GitHub Actions, issue templates)

## Checklist

- [x] My code follows the code style of this project and passes `yarn run lint`.
- [x] My change has adequate [Unit Test coverage](https://publish.obsidian.md/tasks-contributing/Testing/About+Testing).

## Terms

<!--
By submitting this pull request, you must agree to follow our
[contributing guide](https://publish.obsidian.md/tasks-contributing) and
[Code of Conduct](https://github.com/obsidian-tasks-group/obsidian-tasks/blob/main/CODE_OF_CONDUCT.md).
Put an x in the boxes to confirm you agree.
-->

- [x] My contribution follow this project's [contributing guide](https://publish.obsidian.md/tasks-contributing)
- [x] I agree to follow this project's [Code of Conduct](https://github.com/obsidian-tasks-group/obsidian-tasks/blob/main/CODE_OF_CONDUCT.md)

claremacrae commented:
Hi @ilandikov, thanks very much!

I'll look at this as soon as I can...

claremacrae commented on src/lib/QueryRendererHelper.ts:
```diff
@@ -32,10 +31,9 @@ export function explainResults(source: string): string {
-        if (globalQuery.source.trim() !== '') {
-            result += `Explanation of the global query:\n\n${globalQuery.explainQuery()}\n`;
+        const globalQuery = GlobalQuery.getInstance();
+        if (!globalQuery.isEmpty()) {
```
This is also a change in behaviour.

I saw that the previous code trimmed the globalQuery text, and the new code does not trim.

So I tried setting the Global Query to 10 spaces.

I then tried this query:

~~~
```tasks
explain
limit 10
```
~~~

In Tasks 4.6.0, the explanation is:

```text
Only tasks containing the global filter '#task'.

Explanation of this Tasks code block query:

No filters supplied. All tasks will match the query.

At most 10 tasks.
```

With the code in this PR, the explanation is:

```text
Only tasks containing the global filter '#task'.

Explanation of the global query:

No filters supplied. All tasks will match the query.
Explanation of this Tasks code block query:

No filters supplied. All tasks will match the query.

At most 10 tasks.
```

ilandikov replied:
Fixed with a test

claremacrae replied:
I can't see the test, but I tested this out by building the plugin and running it... This confirmed that the behaviour is fixed. Thanks.

claremacrae commented on src/lib/QueryRendererHelper.ts:
```diff
@@ -32,10 +31,9 @@ export function explainResults(source: string): string {
-
-        if (globalQuery.source.trim() !== '') {
-            result += `Explanation of the global query:\n\n${globalQuery.explainQuery()}\n`;
+        const globalQuery = GlobalQuery.getInstance();
```
By using `getInstance()` here and in `explainResults()`, it ties all uses of this code to the global instance, which loses the purpose of this refactoring.

I think this is what must have been unclear from my explanation earlier ... and which in turn is why you were unhappy with the number of calls to `getInstance()`...

Instead, try changing this function so that it takes a GlobalQuery **parameter**.

Now, the code can be supplied a **GlobalQuery** object with any data in it, without needing to change the settings in order to test it...

```diff
Index: src/lib/QueryRendererHelper.ts
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/lib/QueryRendererHelper.ts b/src/lib/QueryRendererHelper.ts
--- a/src/lib/QueryRendererHelper.ts	(revision a697c2273d1d27f3087aec59b8d7b9f04a555cf7)
+++ b/src/lib/QueryRendererHelper.ts	(date 1692127274325)
@@ -1,6 +1,6 @@
-import { GlobalFilter } from '../Config/GlobalFilter';
-import { Query } from '../Query/Query';
-import { GlobalQuery } from '../Config/GlobalQuery';
+import {GlobalFilter} from '../Config/GlobalFilter';
+import {Query} from '../Query/Query';
+import type {GlobalQuery} from '../Config/GlobalQuery';
 
 /**
  * @summary
@@ -20,8 +20,9 @@
  *
  * @param {string} source The source of the task block to explain
  * @returns {string}
+ * @param globalQuery
  */
-export function explainResults(source: string): string {
+export function explainResults(source: string, globalQuery: GlobalQuery): string {
     let result = '';
 
     if (!GlobalFilter.isEmpty()) {
@@ -31,7 +32,6 @@
     const tasksBlockQuery = new Query({ source });
 
     if (!tasksBlockQuery.ignoreGlobalQuery) {
-        const globalQuery = GlobalQuery.getInstance();
         if (!globalQuery.isEmpty()) {
             result += `Explanation of the global query:\n\n${globalQuery.query().explainQuery()}\n`;
         }
@@ -49,14 +49,14 @@
  *
  * @param {string} source The query source from the task block
  * @returns {Query} The query to execute
+ * @param globalQuery
  */
-export function getQueryForQueryRenderer(source: string): Query {
+export function getQueryForQueryRenderer(source: string, globalQuery: GlobalQuery): Query {
     const tasksBlockQuery = new Query({ source });
 
     if (tasksBlockQuery.ignoreGlobalQuery) {
         return tasksBlockQuery;
     }
 
-    const globalQuery = GlobalQuery.getInstance().query();
-    return globalQuery.append(tasksBlockQuery);
+    return globalQuery.query().append(tasksBlockQuery);
 }

```

claremacrae commented on tests/lib/QueryRendererHelper.test.ts:
```diff
@@ -41,7 +40,7 @@ No filters supplied. All tasks will match the query.`;
 
     it('should explain a task with global query active', () => {
-        updateSettings({ globalQuery: 'description includes hello' });
+        GlobalQuery.getInstance().set('description includes hello');
```
The only uses of `GlobalQuery.getInstance()` in tests should be in `GlobalQuery.test.ts`, to test the behaviour of `GlobalQuery.getInstance()`. For example, it might test that calling `GlobalQuery.getInstance()` twice gets back the same instance.

Other than that, all tests should use `new GlobalQuery()` - and then setting the query is only changing local data, not global data.

One thing I noticed when trying this approach out - see below... - is that this test was not doing any resetting of the GlobalQuery to tidy up after itself...

Now all the data is local, that isn't needed any more.

```diff
Index: tests/lib/QueryRendererHelper.test.ts
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/tests/lib/QueryRendererHelper.test.ts b/tests/lib/QueryRendererHelper.test.ts
--- a/tests/lib/QueryRendererHelper.test.ts	(revision a697c2273d1d27f3087aec59b8d7b9f04a555cf7)
+++ b/tests/lib/QueryRendererHelper.test.ts	(date 1692127942347)
@@ -22,7 +22,7 @@
 
 No filters supplied. All tasks will match the query.`;
 
-        expect(explainResults(query.source)).toEqual(expectedDisplayText);
+        expect(explainResults(query.source, new GlobalQuery())).toEqual(expectedDisplayText);
     });
 
     it('should explain a task with global filter active', () => {
@@ -36,11 +36,12 @@
 Explanation of this Tasks code block query:
 
 No filters supplied. All tasks will match the query.`;
-        expect(explainResults(query.source)).toEqual(expectedDisplayText);
+        expect(explainResults(query.source, new GlobalQuery())).toEqual(expectedDisplayText);
     });
 
     it('should explain a task with global query active', () => {
-        GlobalQuery.getInstance().set('description includes hello');
+        const globalQuery = new GlobalQuery();
+        globalQuery.set('description includes hello');
 
         const source = '';
         const query = new Query({ source });
@@ -53,11 +54,12 @@
 
 No filters supplied. All tasks will match the query.`;
 
-        expect(explainResults(query.source)).toEqual(expectedDisplayText);
+        expect(explainResults(query.source, globalQuery)).toEqual(expectedDisplayText);
     });
 
     it('should explain a task with global query and global filter active', () => {
-        GlobalQuery.getInstance().set('description includes hello');
+        const globalQuery = new GlobalQuery();
+        globalQuery.set('description includes hello');
         GlobalFilter.set('#task');
 
         const source = '';
@@ -73,11 +75,12 @@
 
 No filters supplied. All tasks will match the query.`;
 
-        expect(explainResults(query.source)).toEqual(expectedDisplayText);
+        expect(explainResults(query.source, globalQuery)).toEqual(expectedDisplayText);
     });
 
     it('should explain a task with global query set but ignored without the global query', () => {
-        GlobalQuery.getInstance().set('description includes hello');
+        const globalQuery = new GlobalQuery();
+        globalQuery.set('description includes hello');
 
         const source = 'ignore global query';
         const query = new Query({ source });
@@ -86,7 +89,7 @@
 
 No filters supplied. All tasks will match the query.`;
 
-        expect(explainResults(query.source)).toEqual(expectedDisplayText);
+        expect(explainResults(query.source, globalQuery)).toEqual(expectedDisplayText);
     });
 });
 
@@ -99,14 +102,18 @@
     it('should be the result of combining the global query and the actual query', () => {
         const querySource = 'description includes world';
         const globalQuerySource = 'description includes hello';
-        GlobalQuery.getInstance().set(globalQuerySource);
-        expect(getQueryForQueryRenderer(querySource).source).toEqual(`${globalQuerySource}\n${querySource}`);
+        const globalQuery = new GlobalQuery();
+        globalQuery.set(globalQuerySource);
+        expect(getQueryForQueryRenderer(querySource, globalQuery).source).toEqual(
+            `${globalQuerySource}\n${querySource}`,
+        );
     });
 
     it('should ignore the global query if "ignore global query" is set', () => {
-        GlobalQuery.getInstance().set('path includes from_global_query');
-        expect(getQueryForQueryRenderer('description includes from_block_query\nignore global query').source).toEqual(
-            'description includes from_block_query\nignore global query',
-        );
+        const globalQuery = new GlobalQuery();
+        globalQuery.set('path includes from_global_query');
+        expect(
+            getQueryForQueryRenderer('description includes from_block_query\nignore global query', globalQuery).source,
+        ).toEqual('description includes from_block_query\nignore global query');
     });
 });
```

claremacrae commented on src/Config/GlobalQuery.ts:
```diff
@@ -0,0 +1,32 @@
+    }
+
+    public isEmpty() {
+        return this._value === GlobalQuery.empty;
```
I touch on this below, but it would be good to see a new file `GlobalQuery.test.ts` that tests the logic of this class.

For example, I am not really sure of the value of `GlobalQuery.empty`, but the addition of the `isEmpty()` method that does not trim the query has led to an unwanted change in behaviour....

Having discovered this, it would be good to see tests of even this seemingly-very-simple logic, as it has consequences for the behaviour that the user sees...

claremacrae replied:
Somewhere there needs to also be code that tests how the settings are now saved and loaded...

claremacrae reviewed (changes requested):
Hi Ilyas, thanks.

I spent 2 or 3 hours reviewing this.

In view of the behaviour changes, sorry but I don't have time to also review the individual commits unfortunately.

ilandikov commented:
@claremacrae I went through the comments several times, played around with the code and found the piece missing in my head to implement this.

So we don't want to remove the `globalQuery` from `Settings` because the save/load mechanism is working there and extracting the **stored** value somewhere else would be cumbersome - there would be a need to modify the `main.saveSettings()` etc. So we keep the `Settings` as they are and I agree with that 100% now =)

In my understanding the purpose of singleton pattern is to have a unique object that stores the actual data **just once.** But in our case it is the `Settings.settings` that does just that. So `GlobalQuery` is not synthetic sugar but data duplication. So what is the point?

In the previous PR you mentioned that `StatusRegistry` is also using the singleton pattern and if I check that I see that the `statusTypes` are actually stored twice - in the `settings` and in the `StatusRegistry` (with a `static` method that the singleton patter shall remove...):

```ts
async function updateAndSaveStatusSettings(statusTypes: StatusSettings, settings: SettingsTab) {
    updateSettings({
        statusSettings: statusTypes,
    });

    // Update the active statuses.
    // This saves the user from having to restart Obsidian in order to apply the changed status(es).
    StatusSettings.applyToStatusRegistry(statusTypes, StatusRegistry.getInstance());

    await settings.saveSettings(true);
}
```

PS. I'm quite sure that the solution you are proposing is better that what I have in my head, please help me understand it better.

ilandikov commented:
@claremacrae as discussed yesterday

- added tests for GlobalQuery
- fixed the missing `trim()` with TDD ;)
- test are calling `new GlobalQuery()` and src calls `getInstance()`
- fixed storing & restoring the GlobalQuery from the settings. Smoke test:

````
```tasks
description include task block
explain
```
````

set GlobalQuery in settings to `tag includes #globalquery`

restart Obsidan -> explain shows GlobalQuery taken into account.

set GlobalQuery in settings to `tag includes #new_globalquery`

restart Obsidan -> explain shows GlobalQuery taken into account.

ilandikov commented on src/Config/Settings.ts:
```diff
@@ -50,7 +50,6 @@ export const TASK_FORMATS = {
 export type TASK_FORMATS = typeof TASK_FORMATS; // For convenience to make some typing easier
 
 export interface Settings {
-    globalQuery: string;
```
Fixed. `SettingsTab` now acts on both `Settings` and `GlobalQuery`

claremacrae commented on src/Config/GlobalQuery.ts:
```diff
@@ -0,0 +1,32 @@
+import { Query } from '../Query/Query';
+
+export class GlobalQuery {
```
Please document via JSDoc the use of this class - which to use `getInstance()` vs `new ...`.

(There is text in `StatusRegistry` or similar that may be useful).

ilandikov replied:
done =)

claremacrae commented on src/Config/GlobalQuery.ts:
```diff
@@ -0,0 +1,32 @@
+        return new Query({ source: this._source }, path);
+    }
+
+    public isEmpty() {
```
It's highly non-obvious that whitespace is ignored in this calculation - it would be more accurately named as `isEmptyOrPurelyWhitespace()`.

Failing that, please add JSDoc cautioning about this behaviour.

claremacrae replied:
I do appreciate that the trim arose from a previous code review. So just documenting the behaviour is sufficient for now.

ilandikov replied:
Indeed, this should be renamed and the method ignores not just whitespaces but also line breaks. So I gave it some thought, added jsdoc and renamed for it's intent - at the end of the day what this method says is whether the Global Query has any instructions so...- `hasInstructions()`.

This way the tests became much more descriptive:

```js
it.each(['', ' ', '\n', '\n     \n    ', '  \n    \n'])(
        'should have no instructions if line breaks and spaces were set in the query',
        (globalQuerySource) => {
            const globalQuery = new GlobalQuery(globalQuerySource);

            expect(globalQuery.hasInstructions()).toEqual(false);
        },
    );
```

Hope you share this point of view.

claremacrae replied:
> at the end of the day what this method says is whether the Global Query has any instructions so...- hasInstructions()

I like the new name. Thanks.

claremacrae commented on tests/Config/GlobalQuery.test.ts:
```diff
@@ -0,0 +1,33 @@
+    });
+
+    it('should have an empty source by default', () => {
+        expect(new GlobalQuery().isEmpty()).toEqual(true);
```
I would suggest also getting the query and confirming it is an empty string, just for safety.

ilandikov replied:
oh right! done =)

claremacrae commented on tests/Config/GlobalQuery.test.ts:
```diff
@@ -0,0 +1,33 @@
+    });
+
+    it.each(['', ' ', '\n', '\n     \n    ', '  \n    \n'])(
+        'should have empty source if line breaks and spaces were set in the query',
```
I think 'should have empty source' is a bit ambiguous.
Perhaps 'should report that the source is empty'...

ilandikov replied:
Rephrased after `isEmpty()` -> `hasInstructions()` rename.

claremacrae reviewed (approved):
The diffs are lovely and small, especially after such a large marge!

Just a few small suggestions.

claremacrae commented:
> ## How has this been tested?
> Unit tests.

Please also:

- run through the Smoke tests with a build of the latest code
- re-test it manually in the sample vault, confirming that the global query is saved between sessions.

claremacrae commented:
FYI I've tweaked the wording of the title and the initial description to hopefully clarify it a bit, and make it more consistent with the current behaviour.

ilandikov commented:
Manual Global Query test on the latest build:

New file created:
````md
```tasks
```

- [ ] #task #test blabla
````

the query shows tasks from the whole vault.

Set Global Query to `tags include test`, close & open Obsidian, query returns `- [ ] #task #test blabla`.

Close & open Obsidian, query returns the same thing.

Clear Global Query,  close & open Obsidian, query returns tasks from the whole vault.

ilandikov commented:
All smoke tests are passing except this one, however, it is failing also on the current main (the tag is not hidden):

![Снимок экрана 2023-09-22 в 13 23 53](https://github.com/obsidian-tasks-group/obsidian-tasks/assets/93825870/c80ed34d-09b9-4f53-a133-9ae238a6c3db)

claremacrae commented:
> All smoke tests are passing except this one, however, it is failing also on the current main (the tag is not hidden):
> 
> ![Снимок экрана 2023-09-22 в 13 23 53](https://user-images.githubusercontent.com/93825870/269863638-c80ed34d-09b9-4f53-a133-9ae238a6c3db.png)
> 
> Although the `hide tags` works in my private vault...weird...

You had that problem before. It was because you hadn't built the latest version of the plugin in the sample vault.

So it means you aren't smoke-testing the actual behaviour of the changed code in your repo.

I need to streamline that.

I suggest downloading the built plugin from the last build of your branch on GitHub.

claremacrae commented:
Great stuff. Thanks very much.
